### PR TITLE
Improvements to CIBA flow

### PR DESCRIPTION
**Pull request by KalanaDananjaya**
### Proposed changes in this pull request

Public PR for changes to improve CIBA flow suggested in https://github.com/wso2-support/identity-inbound-auth-oauth/pull/992

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/CibaAuthRequestValidator.java [hunk 1]
```diff
@@ -470,13 +467,47 @@ public void validateClient(String request) throws CibaAuthFailureException {
...
+            }
+
+            if (!clientId.equals(authenticatedClient)) {
+                throw new CibaAuthFailureException(OAuth2ErrorCodes.INVALID_CLIENT,
+                        "Issuer does not match with the authenticated client.");
+            }
+            validateClient(request);
+
+        } catch (ParseException ex) {
+            throw new CibaAuthFailureException(OAuth2ErrorCodes.SERVER_ERROR, "Exception in validating for (iss). ");
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/CibaAuthRequestValidator.java:507]: 'ex' not logged or rethrown

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java [hunk 2]
```diff
@@ -192,18 +220,76 @@ private void generateAuthorizeCall(@Context HttpServletRequest request,
...
+
+            OIDCRequestObjectUtil.validateRequestObjectSignature(parameters, requestObject, requestObjectValidator);
+
+            if (!requestObjectValidator.validateRequestObject(requestObject, parameters)) {
+                throw new CibaAuthFailureException(OAuth2ErrorCodes.INVALID_REQUEST, "Invalid parameters " +
+                        "found in the Request Object.");
+
+            }
+        } catch (InvalidRequestException | RequestObjectException e) {
+            throw new CibaAuthFailureException(OAuth2ErrorCodes.INVALID_REQUEST, e.getMessage());
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java:265]: log 'e' as a debug log

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java [hunk 3]
```diff
@@ -192,18 +220,76 @@ private void generateAuthorizeCall(@Context HttpServletRequest request,
...
+        return oAuthClientAuthnContext;
+    }
+
+    private String getSpTenantDomain(String clientId) throws InvalidRequestException {
+
+        try {
+            // At this point we have verified that a valid app exists for the client_id. So we directly get the SP
+            // tenantDomain.
+            return OAuth2Util.getTenantDomainOfOauthApp(clientId);
+        } catch (InvalidOAuthClientException | IdentityOAuth2Exception e) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java:289]: log 'e' as a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+     * @param signedJWT signed JWT
+     * @param x509Certificate X509 certificate
+     * @return signature validity
+     */
+    private static boolean isSignatureVerified(SignedJWT signedJWT, Certificate x509Certificate) {
+
+        JWSVerifier verifier;
+        JWSHeader header = signedJWT.getHeader();
+        if (x509Certificate == null) {
+            return logAndReturnFalse("Unable to locate certificate for JWT " + header.toString());
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java:172]: let's not have a separate method for this. check if debug enabled, then log with the concat, then return false. when passing to a seperate method string concat is already done and no point in checking for is debug enabled. 

Also it's not nice to merge these two which are two independent things

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            } else {
+                return logAndReturnFalse("Public key is not an RSA public key.");
+            }
+        } else {
+            return logAndReturnFalse("Signature Algorithm not supported yet : " + alg);
+        }
+        // At this point 'verifier' will never be null;
+        try {
+            return signedJWT.verify(verifier);
+        } catch (JOSEException e) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java:193]: log 'e'

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 6]
```diff
@@ -1047,6 +1051,31 @@ public RequestObjectValidator getRequestObjectValidator() {
...
+    public RequestObjectValidator getCIBARequestObjectValidator() {
+
+        if (cibaRequestObjectValidator == null) {
+            synchronized (RequestObjectValidator.class) {
+                if (cibaRequestObjectValidator == null) {
+                    try {
+                        Class clazz = Thread.currentThread().getContextClassLoader()
+                                        .loadClass(defaultCibaRequestValidatorClassName);
+                        cibaRequestObjectValidator = (RequestObjectValidator) clazz.newInstance();
+                    } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:1068]: log 'e'

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/CIBARequestObjectValidatorImpl.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import org.wso2.carbon.identity.oauth.RequestObjectValidatorUtil;
+import org.wso2.carbon.identity.oauth2.RequestObjectException;
+import org.wso2.carbon.identity.oauth2.model.OAuth2Parameters;
+import org.wso2.carbon.identity.openidconnect.model.RequestObject;
+
+/**
+ * This class validates request object parameter value which comes with the OIDC CIBA authorization request
+ */
+
+public class CIBARequestObjectValidatorImpl implements RequestObjectValidator {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/CIBARequestObjectValidatorImpl.java:31]: blank line after class sig

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 8]
```diff
@@ -332,92 +227,8 @@ protected boolean validateAudience(String currentAudience, List<String> audience
...
-    }
-
-    /**
-     * Validate the signedJWT signature with given certificate
-     *
-     * @param signedJWT
-     * @param x509Certificate
-     * @return
-     */
-    protected boolean isSignatureVerified(SignedJWT signedJWT, Certificate x509Certificate) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:379]: can't remove protected methods, you can only deprecate

Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:379]: These are moved to `org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil` as discussed keep the common functions of `RequestObjectValidator` and `CIBARequestObjectValidator` in the same place

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/CIBARequestObjectValidatorImpl.java (hunk 7 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/CIBARequestObjectValidatorImpl.java:31]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 6 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:1068]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java (hunk 5 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java:202]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java (hunk 4 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java (hunk 3 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java:292]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java (hunk 2 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java:268]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/CibaAuthRequestValidator.java (hunk 1 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/CibaAuthRequestValidator.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/0d57e5040acf37b8685893a5c5cbd39f29ea9bed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
+package org.wso2.carbon.identity.oauth;
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java:1]: missing licence

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java (hunk 8 above)
Comment (pulasthi7) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:379]: still you can't change any public/protected APIs from last released version of the product. You can only deprecate them and use the newer method underlying

Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:379]: Readded and deprecated with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/01427042989ea11ebcfc9313d4ebf007aac7a19b

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java (hunk 9 above)
Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/RequestObjectValidatorUtil.java:19]: Added with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1743/commits/01427042989ea11ebcfc9313d4ebf007aac7a19b

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 10]
```diff
@@ -231,7 +144,9 @@ private boolean checkExpirationTime(RequestObject requestObject) throws RequestO
                         ", Expiration Time(ms) : " + expirationTimeInMillis +
                         ", TimeStamp Skew : " + timeStampSkewMillis +
                         ", Current Time : " + currentTimeInMillis + ". Token Rejected.";
-                logAndReturnFalse(msg);
+                if (log.isDebugEnabled()) {
+                    log.debug(msg);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:148]: 'msg' should be built inside this block only when debug enabled

Comment (KalanaDananjaya) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java:148]: this is fixed

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**